File: c++/Solution49.cpp
//O(nk) O(nZ), n = strs.size(), k為最大的str長度, Z = 26
class Solution49 {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        unordered_map<string, vector<string>> memo;

        for(string str : strs){
            string hash(26, 0);
            for(char ch : str)
                ++hash[ch - 'a'];

            memo[hash].push_back(str);
        }

        vector<vector<string>> res;
        for(auto it : memo)
            res.push_back(it.second);

        return res;
    }
};

File: python/Solution49.py
# O(nk) O(nZ), n = strs.size(), k為最大的str長度, Z = 26
class Solution49:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        memo = defaultdict(list)
        for s in strs:
            count = [0] * 26
            for c in s:
                count[ord(c) - ord('a')] += 1

            memo[tuple(count)].append(s) # 透過tuple建立hash key
        
        return list(memo.values())